### ISSUE TITLE
`0 as char` triggers the wrong cast diagnostics

### ISSUE THREAD

**Issue by Veykril** [rust-lang/rust-analyzer#18095]
```
fn f() { 0 as char; } 
```
shows "only `u8` can be cast as `char`, not i32rust-analyzer[E0604](https://doc.rust-lang.org/stable/error_codes/E0604.html)"

@ShoyuVanilla

**Comment by ShoyuVanilla** [comment]
I think that this is because we are calling `InferenceTable::fallback_if_possible` before cast checks, so `0` is considered as `i32` rather than coerced to `u8` as like in rustc.
But now I'm on a vacation till this weekend and don't have a laptop 😢 I'll fix this when I'm back

**Comment by Veykril** [comment]
No rush
